hack/generated/controllers/generic_controller.go

```diff
@@ -0,0 +1,516 @@
+	}
+
+	objWrapper := NewReconcileMetadata(metaObj, log)
+	action, actionFunc, err := gr.DetermineReconcileAction(objWrapper)
+
+	if err != nil {
+		log.Error(err, "error determining reconcile action")
+		gr.Recorder.Event(metaObj, v1.EventTypeWarning, "DetermineReconcileActionError", err.Error())
+	}
+
+	result, err := actionFunc(ctx, action, objWrapper)
+	if err != nil {
+		log.Error(err, "Error during reconcile", "action", action)
+		gr.Recorder.Event(metaObj, v1.EventTypeWarning, "ReconcileActionError", err.Error())
+	}
+
+	return result, err
+}
+
+func (gr *GenericReconciler) DetermineReconcileAction(obj *ReconcileMetadata) (ReconcileAction, ReconcileActionFunc, error) {
+	if !obj.metaObj.GetDeletionTimestamp().IsZero() {
+		if obj.resourceProvisioningState != nil && *obj.resourceProvisioningState == armclient.DeletingProvisioningState {
+			return ReconcileActionWatchDelete, gr.UpdateFromNonTerminalDeleteState, nil
+		}
+		return ReconcileActionBeginDelete, gr.StartDeleteOfResource, nil
+	} else {
+		hasChanged, err := genruntime.HasResourceSpecHashChanged(obj.metaObj)
+		if err != nil {
+			return ReconcileActionNoAction, NoAction, errors.Wrap(err, "failed comparing resource hash")
+		}
+
+		//if !hasChanged && r.IsTerminalProvisioningState() && hasStatus {
+		if !hasChanged && obj.IsTerminalProvisioningState() {
+			// TODO: Do we want to log here?
+			msg := fmt.Sprintf("resource spec has not changed and resource is in terminal state: %q", *obj.resourceProvisioningState)
+			obj.log.V(0).Info(msg)
+			return ReconcileActionNoAction, NoAction, nil
+		}
+
+		if obj.resourceProvisioningState != nil && *obj.resourceProvisioningState == armclient.DeletingProvisioningState {
+			return ReconcileActionNoAction, NoAction, errors.Errorf("resource is currently deleting; it can not be applied")
+		}
+
+		if obj.deploymentId != "" {
+			// There is an ongoing deployment we need to monitor
+			return ReconcileActionWatchDeployment, gr.WatchDeployment, nil
+		}
+
+		return ReconcileActionBeginDeployment, gr.CreateDeployment, nil
+	}
+}
+
+//////////////////////////////////////////
+// Actions
+//////////////////////////////////////////
+
+func NoAction(ctx context.Context, action ReconcileAction, obj *ReconcileMetadata) (ctrl.Result, error) {
+	return ctrl.Result{}, nil
+}
+
+// StartDeleteOfResource will begin the delete of a resource by telling Azure to start deleting it. The resource will be
+// marked with the provisioning state of "Deleting".
+func (gr *GenericReconciler) StartDeleteOfResource(
+	ctx context.Context,
+	action ReconcileAction,
+	obj *ReconcileMetadata) (ctrl.Result, error) {
+
+	msg := "Starting delete of resource"
+	obj.log.Info(msg)
+	gr.Recorder.Event(obj.metaObj, v1.EventTypeNormal, string(action), msg)
+
+	resource, err := gr.constructArmResource(ctx, obj)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "couldn't convert to armResourceSpec")
```

Comment:
Added identifying information.

I still haven't quite gotten a handle on when we need it vs when we don't. If we were trapping and logging all of our errors with our logger we dont need it really because it is already added as a kvp on the logger...

For now we can err on the side of "too much information" and we can always prune it later though.